### ISSUE TITLE
Target's date property with undefined value shouldn't default to today

### ISSUE THREAD

**Issue by latin-panda** [medic/cht-core#7711]
**Describe the bug**
The Provider-wireup's [targetEmissionFilter](https://github.com/medic/cht-core/blob/7c0b599af4f9c6277070ec8cc2b41e6e2b436b8a/shared-libs/rules-engine/src/provider-wireup.js#L143) returns today date when date is `undefined`. 
Reading [the documentation about targets](https://docs.communityhealthtoolkit.org/apps/reference/targets/#targetsjs), the `date` property default value is `reported`, and the value to get a today date is `now`. Therefore,  If the `date` property is undefined, then it should default to `reported`, or at least it should consider that record in the target. As it is right now, it feels like we have 2 defaults values (today date and reported). 

Some projects might be working like this. We should highlight this fix when releasing, so app builders ensures targets are okay. 

**To Reproduce**
1. Create a target like this:
```
// Taken from default config 
{
    id: 'births-this-month',
    type: 'count',
    icon: 'icon-infant',
    goal: -1,
    translation_key: 'targets.births.title',
    subtitle_translation_key: 'targets.this_month.subtitle',
    appliesTo: 'contacts',
    appliesToType: ['person'],
    appliesIf: function (contact) {
      return !!contact;
    },
    date: (contact) => contact.contact.date_of_birth,
    dhis: {
      dataElement: 'kB0ZBFisE0e',
    }
  }
```
3. Upload the config
4. Create a contact without `date_of_birth`
5. See how this contact is counted in "birth this month" when it shouldn't. It doesn't even have a date of birth. 

**Expected behavior**
 If the `date` property is undefined, then it should default to `reported`, or at least it should consider that record in the target.

**Environment**
- Instance: local
- Browser: Chrome
- Client platform: MacOS
- App: shared-libs
- Version: 3.15 (but looks like it's been around [since rules was introduced](https://github.com/medic/cht-core/blob/3.8.x/shared-libs/rules-engine/src/provider-wireup.js#L96)) 

**Additional context**
I've fixed the target in default config and the e2e that was testing this ([see PR](https://github.com/medic/cht-core/pull/7710)). 
My user has a contact associated but without date of birth, it was being counted as a new born in that target.

**Comment by latin-panda** [comment]
@dianabarsan @garethbowen This is the issue I found while debugging a flaky e2e.

We can:
- default to `reported`, as the docs says
- ignore the record because it's unclear how to default it. The result is the equivalent of "appliesIf" result.
- __________ (other)

**Comment by garethbowen** [comment]
Excellent detective work!

I think there's another twist in this saga... if no date property is provided, the [target emitter code](https://github.com/medic/cht-conf/blob/main/src/nools/target-emitter.js#L24-L38) _actually_ defaults to `'now'` not `'reported'`. So we need to decide whether to change the docs to match the code, or the code to match the docs. I haven't tested this but if you wanted it would be good to check that if you just delete the `date` property the behaviour is consistent with if you provide a function that returns `null`.

Because changing the behaviour would be a breaking change I propose we change the docs to specify `'now'` as default and leave the code unchanged as I think it's consistent currently.

Frankly I think using `reported` for the default makes more sense, but I'm cautious about breaking this, and projects can deal with the null case as you have done in the test.

**Comment by garethbowen** [comment]
@latin-panda What do you think about my proposal above?

**Comment by latin-panda** [comment]
@garethbowen I haven't had time to debug this one again 😞 
Defaulting to `reported` makes more sense to me since I assume targets are usually worked around summited reports in a window of time. We can always validate our assumption by running a quick question to app builders about the understanding of the defaults.
I think it can be a quick win together with this other [issue](https://github.com/medic/cht-core/issues/7719).

**Comment by garethbowen** [comment]
I agree it makes more sense but we can't break backwards compatibility without a really good reason.

**Comment by latin-panda** [comment]
> I haven't tested this but if you wanted it would be good to check that if you just delete the date property the behaviour is consistent with if you provide a function that returns null.

@garethbowen I retested this statement, by deleting the date property from the target and re-uploading my config. 
The result is the same if the date property missing as if the date is a function returning `undefined`, both cases are considered as `now`.  In this case, as you suggested we can let it default to `now` and update the documentation  explaning of this behaviour, so it won't be a breaking change.  

If the date is a function returning `null`, then it fails the [targetEmissionFilter](https://github.com/medic/cht-core/blob/7c0b599af4f9c6277070ec8cc2b41e6e2b436b8a/shared-libs/rules-engine/src/provider-wireup.js#L143) check `moment(emission.date).isBetween(filterInterval.start, filterInterval.end, null, '[]')` as emission.date is null, and the target emissions aren't counted. 
Should this escenario also default to `now`? If not then, should we print a warning?

**Comment by garethbowen** [comment]
@latin-panda Interestingly passing `undefined` to moment returns now, but passing `null` returns null. I thought those would be equivalent. I presume this also means if you refer to a property (like reported date) the behaviour will change whether the property is `null` or `undefined`. I'm going to make some assertions - please comment if you disagree!

1. Targets dates should be consistent whether or not the date was the result of a property lookup or a function (a null is a null).
2. Date values of `null` and `undefined` should behave the same

Of these two I think the second one is more controversial, but I think it's unintuitive that `null` and `undefined` would result in different outcomes.

Do you agree with these assertions?

If so, we just need to agree on whether `null` and `undefined` should default to `reported`, `now`, or never match the `targetEmissionFilter`.

**Comment by latin-panda** [comment]
> - Targets dates should be consistent whether or not the date was the result of a property lookup or a function (a null is a null).
> - Date values of null and undefined should behave the same

Agree with both bullets. 

> Of these two I think the second one is more controversial, but I think it's unintuitive that null and undefined would result in different outcomes.

Yes, it's not intuitive.

> If so, we just need to agree on whether null and undefined should default to reported, now, or never match the targetEmissionFilter.

I think `null` and `undefined` can default to `now`, just as the [target emitter code](https://github.com/medic/cht-conf/blob/main/src/nools/target-emitter.js#L24-L38) do. We need to explain in detail in the cht-docs.

**Comment by garethbowen** [comment]
Sounds good.

So, for consistency, `null` and `undefined` should default to "now", regardless of whether a property or function is provided. This applies to the target emitter code and the target emission filter and anywhere else these values are used. Secondly the docs should be updated with "now" as the default.

If we still think the default value should be something different, then we should raise a separate issue for that so it can be communicated as a breaking change.

**Comment by elvisdorkenoo** [comment]
This [PR](https://github.com/medic/cht-conf/pull/514) fixes the issue and is ready for AT.

**Comment by latin-panda** [comment]
I think it'd be better to avoid confusion. Thanks!
